src/main/java/com/mineboundteam/minebound/magic/UtilitySpells/EarthUtilitySpell.java

```diff
@@ -0,0 +1,261 @@
+                int speedLevel = -1;
+                for (EarthUtilitySpell s : equippedSpells) {
+                    manaReduction += s.config.MANA_REDUCTION.get();
+                    speedLevel += s.config.SPEED_LEVEL.get();
+                }
+
+                if(speedLevel > -1){
+                    int finalManaReduction = manaReduction;
+                    player.getCapability(PlayerManaProvider.PLAYER_MANA).ifPresent(playerMana -> playerMana.setManaCapModifier("earth_utility", -finalManaReduction));
+                    player.addEffect(new MobEffectInstance(MobEffects.DIG_SPEED, 2, speedLevel, false, false));
+                }
+            } else {
+                // Remove total mana reduction
+                player.getCapability(PlayerManaProvider.PLAYER_MANA).ifPresent(playerMana -> playerMana.setManaCapModifier("earth_utility", 0));
+            }
+        }
+    }
+
+    protected final static List<Vec3> shifts = new ArrayList<>();
+    protected static boolean doingSearch = false;
+
+    static {
+        for(int i = -1; i <= 1; i++)
+            for(int j = -1; j <= 1; j++)
+                for(int k = -1; k <= 1; k++)
+                    // stupid language
+                    if(!(i == 0 && i == j && j == k))
+                        shifts.add(new Vec3(i, j, k));
+    }
+
+
+    // We want this to run first because it'll cancel and rerun for each selected block
+    // we don't want a bunch of potential subscribe events to run for no reason on the initial break.
+    @SubscribeEvent(priority = EventPriority.HIGHEST)
+    public static void onBreak(BlockEvent.BreakEvent event){
+        // While sending a blockbreak event from any other source, this should trigger.
+        // But, since it breaks blocks too, there needs to be a check to prevent infinite recursion
+        if(!doingSearch && event.getPlayer() instanceof ServerPlayer player && !player.isCreative()){
+            // Check if the player even wants to be vein mining. If not, we can safely leave the method.
+            UtilityToggle toggle = player.getCapability(PlayerUtilityToggleProvider.UTILITY_TOGGLE).resolve().get();
+            if(toggle != null && !toggle.earth)
+                return;            
+
+            EarthUtilitySpell spell = getHighestSpellItem(EarthUtilitySpell.class, player);
+            Level level = player.level; 
+            BlockState block = event.getState();
+
+            // If we're doing tag checking, make sure the block is vein-mineable 
+            if(spell == null || (USE_TAGS.get() && !block.getTags().anyMatch(t -> t.location().equals(VEIN_TAGS))))
+                return;
+
+            // Simple breadth first search; could implement a quicker search using the block type as a heuristic
+            LinkedList<BlockPosSearch> open = new LinkedList<>();
+            LinkedList<BlockPosSearch> closed = new LinkedList<>();
```

Comment:
The only thing I noticed was it will still vein mine blocks that have a 1-block gap between them.

Reply:
That's the tolerance value; I have it set to 2 by default for diamonds